041-opa-integration/proposal.md

```diff
@@ -0,0 +1,149 @@
+# Summary
+
+This proposal outlines the beginnings of support for policy enforcement. OPA 
+(Open Policy Agent) will be used as the first policy manager to integrate with.
+
+
+# Motivation
+
+Generally speaking, anything that exposes an HTTP API (whether an individual 
+micro-service or an application as a whole) needs to control who can run those 
+APIs and when. 
+
+Concourse, as a CI platform, also needs to control who can do something and when,
+which can be done by integrating with a policy manager. Before any action, send
+a policy check request, with data that describes the action, to the policy manager,
+and only continue if the policy manager replies a "pass".
+
+Possible policies to apply could be:
+
+* Steps, e.g. code security scan, by certain resource types must be done.
+* Some team's tasks must be run with certain tag.
+* Some docker image registries are not allowed.
+* Some docker images are not not allowed.
+* and so on ...
+
+
+# Proposal
+
+## Support multiple policy managers
+
+Like how multiple credential managers are supported, Concourse should allow other
+policy managers than OPA. Thus an interface of `PolicyCheck` should be defined, and
+OPA is one of the implementations.
+
+## Policy check points
+
+* All API calls, for example `set-pipeline`, will also go through policy checks. If
+the check doesn't pass, API should return HTTP code 403 (forbidden).
+* `UsingImage` action will be sent to OPA before Concourse launches a container in
+`check/get/put/task` steps. If the check doesn't pass, the step should return an error
+indicating policy check not pass.
+
+## OPA configuration
+
+* CONCOURSE_OPA_URL - URL of OPA service including path
+
+For example, OPA_URL is `http://opa.mycompany.com/v1/data/concourse/policy`, a OPA 
+policy check request will look like:
+
+```
+POST http://opa.mycompany.com/v1/data/concourse/policy
+{
+   "input": {
+      ...
+   }
+}
+```
+
+## Policy check input data
+
+Policy check input data should include common metadata and action specific data. 
+In principle, Concourse should send as much data as possible to the policy engine.
+
+Common metadata include:
+
+* `service`: should be always "concourse". This will allow Concourse specific 
+policies to to configured in the policy engine.
+* `cluster_name`: cluster name.
+* `cluster_version`: Concourse version
+* `action`: Action name. This follows same action names used in RBAC, plus an extra
+action `UsingImage`.
+* `http_method`: HTTP method of the action, for `UsingImage`, this field absents.
+* `user`: username who invokes the action.
+* `team`: team name.
+* `pipeline`: pipeline name. Some action is not against a pipeline, then this field 
+can be omitted.
+* `data`: For API actions, `data` should carry data from the API call; for `UsingImage`
```

Comment:
It's probably worth noting here *what* data gets sent during `UsingImage` - we've gone through a few iterations in https://github.com/concourse/concourse/pull/5034

In particular it's probably worth mentioning how credentials are handled, and the potential risk of transmitting them if a credential manager is not used or if secret redaction is not enabled.

Reply:
Added a note for image_source redaction.